### PR TITLE
Simplify typeinfo.d to take advantage of templated implementation

### PR DESCRIPTION
**Pull request by andralex**
Switching to templates allow us to eliminate `Floating` and `Array` entirely.

### CODE REVIEW

**Comment by Geod24** [comment]
> As usual, please give a proper title to both PR and commits.
> Remember the [Contributing guidelines](https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md#solving-bugs--submitting-pull-requests) include an article about [good commit messages](https://chris.beams.io/posts/git-commit/).

I should write a bot for this.

**Comment by wilzbach** [comment]
> > As usual, please give a proper title to both PR and commits.
> > Remember the [Contributing guidelines](https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md#solving-bugs--submitting-pull-requests) include an article about [good commit messages](https://chris.beams.io/posts/git-commit/).
> 
> I should write a bot for this.

Maybe we should add it to the dlang-bot's message?

**Comment by Geod24** [comment]
> Maybe we should add it to the dlang-bot's message?

Experience shows that everyone just ignores it. Including me. I only look at it to see if the issue will properly auto-close (and since I know the syntax, I rarely do that). Note that an alternative for this would be to use PR templates.
I think it would only really work if the bot could give you real feedback on your PR instead of blindly pasting a message. But that's not easily done with commit messages.

**Comment by andralex** [comment]
Changed the commit msg and the title.

**Comment by andralex** [comment]
It bothers me that `TypeInfoXxx.compare` is just slightly different for floating point numbers than the built-in comparison (enshrined by `__cmp`). I wonder if we should just make all behave like `__cmp`. (I tried it, no unittests failed.)

**Comment by andralex** [comment]
Added to https://issues.dlang.org/show_bug.cgi?id=13499

File: src/rt/util/typeinfo.d [hunk 1]
```diff
@@ -1,107 +1,64 @@
...
-    }
-
-    public alias hashOf = core.internal.hash.hashOf;
+// Three-way compare for real fp types. NaN is smaller than all valid numbers.
+// Code is small and fast, see https://godbolt.org/z/fzb877
+pragma(inline, true)
+private int cmp3(T)(const T d1, const T d2)
+if (is(T == float) || is(T == double) || is(T == real))
+{
+    if (d2 != d2)
```
Comment (andralex) [inline src/rt/util/typeinfo.d:29]: This codecov warning seems spurious - the unittest below exercise all paths in this function. cc @WalterBright

File: src/core/internal/array/comparison.d [hunk 2]
```diff
@@ -60,26 +60,24 @@ int __cmp(T)(scope const T[] lhs, scope const T[] rhs) @trusted
         immutable len = lhs.length <= rhs.length ? lhs.length : rhs.length;
         foreach (const u; 0 .. len)
         {
-            static if (__traits(isFloating, T))
-            {
-                immutable a = lhs.ptr[u], b = rhs.ptr[u];
-                static if (is(T == cfloat) || is(T == cdouble)
+            auto a = lhs.ptr[u], b = rhs.ptr[u];
+            static if (__traits(isFloating, T) && is(T == cfloat) || is(T == cdouble)
```
Comment (n8sh) [inline src/core/internal/array/comparison.d:64]: I suggest replacing the entire condition with the more succinct
```d
static if (is(T : creal))
````
 I have verified that for scalar `T` this has no performance penalty relative to
```d
static if (__traits(isFloating, T) && (is(T == cfloat) || is(T == cdouble) || is(T == creal)))
```
(in fact it is ever-so-slightly faster and ever-so-slightly more memory efficient).

Comment (andralex) [reply src/core/internal/array/comparison.d:64]: OK

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Pease rebase

**Comment by RazvanN7** [comment]
@andralex @thewilsonator I have rebased this and updated it to not use builtin complex numbers (which have been deprecated). Is this good to go?

**Comment by thewilsonator** [comment]
Yes although this PR says This branch cannot be rebased due to conflicts.

**Review by RazvanN7 — APPROVED** [review]

**Comment by andralex** [comment]
@RazvanN7 thanks for taking this over the finish line!

**Pull request merged**